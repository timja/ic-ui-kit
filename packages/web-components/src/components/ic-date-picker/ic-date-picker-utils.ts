import { IcWeekDays } from "../../utils/types";

const IC_ISO_DATE_FORMAT = /^(\d{4})-(\d{2})-(\d{2})$/;

export const createDateFromISOString = (isoDate: string): Date => {
  if (isoDate === null || isoDate === "" || isoDate === undefined) {
    return null;
  }

  const matches = isoDate.match(IC_ISO_DATE_FORMAT);

  if (matches) {
    return createDate(matches[1], matches[2], matches[3]);
  }

  return null;
};

export const createDate = (year: string, month: string, day: string): Date => {
  const dayInt = parseInt(day, 10);
  const monthInt = parseInt(month, 10);
  const yearInt = parseInt(year, 10);

  const isValid =
    Number.isInteger(yearInt) && // all parts should be integers
    Number.isInteger(monthInt) &&
    Number.isInteger(dayInt) &&
    monthInt > 0 && // month must be 1-12
    monthInt <= 12 &&
    dayInt > 0 && // day must be 1-31
    dayInt <= 31 &&
    yearInt > 0;

  if (isValid) {
    return new Date(yearInt, monthInt - 1, dayInt);
  }
  return null;
};

export const getWeekStart = (date: Date, startDay: IcWeekDays): Date => {
  const tmpDate = new Date(date);
  const day = tmpDate.getDay();
  const diff = (day < startDay ? 7 : 0) + day - startDay;

  tmpDate.setDate(tmpDate.getDate() - diff);
  return tmpDate;
};

export const getWeekEnd = (date: Date, startDay: IcWeekDays): Date => {
  const d = new Date(date);
  const day = d.getDay();
  const diff = (day < startDay ? -7 : 0) + 6 - (day - startDay);

  d.setDate(d.getDate() + diff);
  return d;
};

export const getMonthStart = (date: Date): Date => {
  return new Date(date.getFullYear(), date.getMonth(), 1);
};

export const getMonthEnd = (date: Date): Date => {
  return new Date(date.getFullYear(), date.getMonth() + 1, 0);
};

/**
 * Compare if two dates are equal exactly equal
 */
export const dateMatches = (a: Date, b: Date): boolean => {
  if (a === null || b === null) {
    return false;
  }

  return (
    a.getFullYear() === b.getFullYear() &&
    a.getMonth() === b.getMonth() &&
    a.getDate() === b.getDate()
  );
};

export const clampDate = (date: Date, min: Date, max: Date): Date => {
  const time = date.getTime();
  if (min !== null && time < min.getTime()) {
    return min;
  }
  if (max !== null && time > max.getTime()) {
    return max;
  }
  return date;
};

/**
 * Compare if date is within specified range
 */
export const dateInRange = (date: Date, min: Date, max: Date): boolean => {
  return clampDate(date, min, max) === date;
  // if (min === null && max === null) {
  //   return true;
  // } else if (min !== null && max !== null) {
  //   return date >= min && date <= max;
  // } else if (min !== null) {
  //   return date >= min;
  // } else {
  //   return date <= max;
  // }
};

export const yearInRange = (year: number, min: Date, max: Date): boolean => {
  let allowed = true;
  if (year !== null) {
    if (min !== null && year < min.getFullYear()) {
      allowed = false;
    }
    if (allowed && max !== null && year > max.getFullYear()) {
      allowed = false;
    }
  }
  return allowed;
};
