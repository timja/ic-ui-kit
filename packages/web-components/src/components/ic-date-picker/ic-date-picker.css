@import "../../global/normalize.css";

/**
* @prop --input-width: Width of the input field
*/

:host {
  /* z-index: var(--ic-z-index-popover); */
  display: block;
  position: relative;

  --date-picker-flex-gap: var(--ic-space-xxs);
  --month-year-picker-button-width: 6.5rem;
  --month-button-width: 5.5rem;
  --input-field-width: var(--input-width, 314px);
}

:host(.large) {
  --date-picker-flex-gap: var(--ic-space-xs);
  --month-button-width: 6.5rem;
  --input-field-width: var(--input-width, 354px);
}

:host(.small) {
  --month-year-picker-button-width: 5rem;
  --month-button-width: 6.5rem;
  --input-field-width: var(--input-width, 266px);
}

:host .date-input-container {
  position: relative;
}

ic-date-input {
  --input-width: var(--input-field-width);
}

.hidden-describe {
  display: none;
}

.sr-only {
  position: absolute;
  left: -9999px;
}

.calendar-button {
  position: absolute;
  left: 336px;
  bottom: 4px;
}

:host .calendar-container {
  /* max-height: 0; */
  height: 0;

  /* display: none; */
  display: flex;
  flex-direction: column;
  gap: var(--date-picker-flex-gap);
  visibility: hidden;

  /* height: 0; */
  position: absolute;
  width: 19.625rem;
  border: var(--ic-border-default);
  border-radius: var(--ic-border-radius);
  align-items: center;
  background-color: var(--ic-architectural-white);
  z-index: var(--ic-z-index-date-picker);
  box-sizing: border-box;
  box-shadow: var(--ic-elevation-overlay);
  margin-top: var(--ic-space-xs);
  padding: var(--ic-space-xs) var(--ic-space-xxs);

  /* transition: height var(--ic-transition-duration-slow); */

  /* animation: fadeOutCalendar var(--ic-transition-duration-slow); */

  /* animation: fadeOutCalendar 1000ms; */
  opacity: 0;
}

:host(.large) .calendar-container {
  width: 22.125rem;
  padding: var(--ic-space-xs);
}

:host(.small) .calendar-container {
  width: 16.625rem;
  padding: var(--ic-space-xxxs) var(--ic-space-xxs);
}

.calendar-container.open {
  opacity: 1;
  visibility: visible;
  height: auto;
  transition: opacity var(--ic-transition-duration-slow);

  /* transition: none; */

  /* animation: fadeInCalendar var(--ic-transition-duration-slow); */
}

:host .month-year-nav-container {
  display: flex;
  justify-content: space-between;
  align-items: center;
  align-self: stretch;
}

:host .month-year-nav-container.hidden {
  display: none;
}

:host .month-year-nav {
  display: flex;
  justify-content: center;
  align-items: center;
}

:host .month-picker {
  flex-wrap: wrap;
  display: inline-flex;
  padding: var(--ic-space-xs) var(--ic-space-md);
  align-items: center;
  row-gap: var(--date-picker-flex-gap);
  column-gap: var(--ic-space-xxxs);
}

:host(.small) .month-picker {
  column-gap: var(--ic-space-md);
  row-gap: var(--ic-space-xxs);
  padding: var(--ic-space-xxs) var(--ic-space-md);
}

:host(.large) .month-picker {
  padding: var(--ic-space-xs) var(--ic-space-xxs);
  column-gap: var(--ic-space-xs);
  row-gap: var(--ic-space-xs);
}

:host .month-picker-button,
:host .year-picker-button {
  width: var(--month-year-picker-button-width);

  --min-width: var(--month-year-picker-button-width);
}

:host .month-button,
:host .year-button,
:host .decade-button {
  width: var(--month-button-width);
}

:host .month-button::part(button),
:host .year-button::part(button),
:host .decade-button::part(button) {
  border-radius: 24px;
  min-width: var(--month-button-width);
}

:host .month-button.disabled::part(button),
:host .year-button.disabled::part(button) {
  text-decoration-line: line-through;
}

:host .year-picker-container {
  flex-wrap: wrap;
  display: inline-flex;
  padding: var(--ic-space-xs) var(--ic-space-xs) var(--ic-space-xs)
    var(--ic-space-sm);
  align-items: center;
  column-gap: var(--ic-space-xs);
  row-gap: var(--ic-space-xs);
}

:host(.small) .year-picker-container {
  padding: var(--ic-space-xs) var(--ic-space-md) var(--ic-space-xs)
    var(--ic-space-lg);
}

:host(.large) .year-picker-container {
  padding: var(--ic-space-xs) var(--ic-space-xxs);
}

:host .bottom-buttons {
  padding-top: var(--ic-space-md);
  display: flex;
  justify-content: space-between;
  align-items: center;
  align-self: stretch;
}

:host .bottom-buttons.no-today {
  align-items: flex-end;
  flex-direction: column;
}

:host .bottom-buttons.hidden {
  display: none;
}

:host .calendar {
  display: flex;
  flex-direction: column;
  padding: 0.625rem 1.75rem;
  align-items: flex-start;
  width: 15.75rem;
}

:host(.large) .calendar {
  padding: 0 1rem;
  align-items: center;
  width: 19.25rem;
}

:host(.small) .calendar {
  padding: 0.625rem 0;
  width: 14rem;
}

:host .hidden {
  display: none;
}

:host .weekdays {
  display: flex;
  justify-content: space-between;
  align-items: flex-start;
  width: 15.75rem;
  padding: 0 var(--ic-space-xxxs) var(--ic-space-xs) 0;
  flex: 1 0 0;
}

:host(.small) .weekdays {
  width: 13.75rem;
  padding: 0 var(--ic-space-xxxs) var(--ic-space-xs) var(--ic-space-xxxs);
}

:host(.large) .weekdays {
  width: 19.25rem;
  padding: 0 var(--ic-space-xxs) var(--ic-space-md) var(--ic-space-xxs);
}

:host .calendar-days-container {
  display: flex;
  flex-wrap: wrap;
  column-gap: 0;
  row-gap: var(--ic-space-md);
  padding: var(--ic-space-xs) 0;
}

:host(.small) .calendar-days-container {
  row-gap: var(--ic-space-xs);
  padding: var(--ic-space-xxs) 0;
}

:host .calendar-day-header {
  display: flex;
  width: 2.25rem;
  padding: var(--ic-space-xxs);
  justify-content: center;
  align-items: center;
  gap: 0.625rem;
}

:host(.small) .calendar-day-header {
  width: 1.75rem;
  padding: var(--ic-space-xxs) 0;
  justify-content: space-between;
  gap: 0;
}

:host .calendar-day-header ic-typography {
  color: var(--ic-color-tertiary-text);
  text-align: center;
  display: flex;
  width: 1.75rem;
  flex-direction: column;
  flex-shrink: 0;
}

/* @media (prefers-reduced-motion: no-preference) { */

/* .calendar-container {
    animation: fadeOutCalendar var(--ic-transition-duration-slow);
  } */

/* 
  .calendar-container.open {
    animation: fadeInCalendar var(--ic-transition-duration-slow);
  }

  @keyframes fadeInCalendar {
    0% {
        display: none;
        opacity: 0;
    }

    1% {
        display: block;
        opacity: 0;
    }

    100% {
        display: block;
        opacity: 1;
    }
  }

  @keyframes fadeOutCalendar {
    0% {
        display: block;
        opacity: 1;
    }

    1% {
        display: block;
        opacity: 0;
    }

    100% {
        display: none;
        opacity: 0;
    }
  }

} */

/* @keyframes fadeInCalendar {
  0% {
      display: none;
      opacity: 0;
  }

  1% {
      display: block;
      opacity: 0;
  }

  100% {
      display: block;
      opacity: 1;
  }
}

@keyframes fadeOutCalendar {
  0% {
      display: block;
      opacity: 1;
  }

  1% {
      display: block;
      opacity: 0;
  }

  100% {
      display: none;
      opacity: 0;
  }
} */

:host .day-button {
  display: flex;
  position: relative;
  justify-content: center;
  align-items: center;
  border: 0;
  padding: 0;
  border-radius: 2rem;
  background-color: var(--ic-architectural-white);
  width: 1.75rem;
  height: 1.75rem;
  margin: 0 4px;
  cursor: pointer;
}

:host .day-button.disabled {
  cursor: default;
}

:host(.large) .day-button {
  width: 2rem;
  height: 2rem;
  margin: 0 6px;
}

:host(.small) .day-button {
  margin: 0 2px;
}

:host .day-button.outside-range ic-typography {
  color: var(--ic-architectural-400);
}

:host .day-button.outside-month ic-typography {
  --custom-font-weight: 400;
}

:host
  .day-button.outside-month:not(.outside-range):not(.selected)
  ic-typography {
  color: var(--ic-color-primary-text);
}

:host .day-button:hover:not(.disabled) {
  background-color: var(--ic-action-dark-bg-hover);
  transition: all var(--ic-transition-duration-slow) ease-in-out;
}

:host .day-button:active:not(.disabled) {
  background-color: var(--ic-action-dark-bg-active);
  transition: all var(--ic-transition-duration-slow) ease-in-out;
}

:host .day-button.selected {
  background-color: var(--ic-action-default);
}

:host .day-button.selected ic-typography {
  color: var(--ic-color-white-text);
}

:host .day-button.selected:hover {
  background-color: var(--ic-action-default-hover);
}

:host .day-button.selected:active {
  background-color: var(--ic-action-default-active);
}

:host .day-button:focus {
  outline: none;
}

:host .day-button.focussed:focus {
  box-shadow: var(--ic-border-focus);
  z-index: 2;
}

:host .day-button.today::after {
  content: "";
  position: absolute;
  width: 0.875rem;
  height: 0.125rem;
  top: 1.375rem;
  border-radius: 2px;
  background-color: var(--ic-action-default);
}

:host(.large) .day-button.today::after {
  width: 1rem;
  top: 1.625rem;
}

:host .day-button.today.selected::after {
  background-color: var(--ic-architectural-white);
}

:host #select-month-hint,
:host #select-year-hint {
  display: none;
}

/* @media (prefers-reduced-motion: no-preference) {
  .calendar-container.open {
    
  }
} */

/*
:host {
  border-radius: var(--ic-border-radius);
  color: var(--ic-color-primary-text);
  background-color: var(--ic-architectural-white);
  position: relative;
  z-index: var(--ic-z-index-popover);
  box-sizing: border-box;
  box-shadow: var(--ic-elevation-overlay);
  display: none;
}

.menu {
  border: 0.063rem solid var(--ic-architectural-400);
  border-radius: var(--ic-border-radius);
  background-color: var(--ic-architectural-white);
  visibility: hidden;
  height: 0;
}

.button {
  text-decoration: none;
  list-style-type: none;
}

:host(:focus-within) {
  box-shadow: var(--ic-border-focus);
}

.menu:focus-visible {
  outline: none;
}

:host(.open) {
  display: block;
  min-width: calc(20rem - var(--ic-space-xl));
  width: var(--popover-width, 20rem);
  max-width: calc(100vw - var(--ic-space-xl));
}

:host(.open) .menu {
  visibility: visible;
  height: fit-content;
  max-height: var(--max-height, fit-content);
  overflow-y: auto;
  overflow-x: hidden;
}

.parent-label {
  color: var(--ic-color-tertiary-text);
  margin: var(--ic-space-xs) var(--ic-space-xs) 0;
}

@keyframes slide-in {
  from {
    opacity: 0;
    transform: translateX(10rem);
  }

  to {
    opacity: 1;
    transform: translateX(0);
  }
}

@keyframes slide-out {
  from {
    opacity: 0;
    transform: translateX(-10rem);
  }

  to {
    opacity: 1;
    transform: translateX(0);
  }
}
*/
