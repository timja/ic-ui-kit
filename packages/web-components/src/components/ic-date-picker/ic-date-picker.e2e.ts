import { newE2EPage } from "@stencil/core/testing";

describe("ic-date-picker component", () => {
  it("renders", async () => {
    const page = await newE2EPage();
    await page.setContent(`<ic-date-picker></ic-date-picker>`);

    const element = await page.find("ic-date-picker");

    expect(element).toHaveClass("hydrated");
  });
});
