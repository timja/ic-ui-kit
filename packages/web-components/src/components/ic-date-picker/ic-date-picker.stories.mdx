import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Description,
} from "@storybook/addon-docs";
import { html } from "lit-html";
import readme from "./readme.md";

<Meta title="Web Components/Date picker" component="Date Picker" />

<Description markdown={readme} />

### Default

<Canvas>
  <Story name="Default" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Large

<Canvas>
  <Story name="Large" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          size="large"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Small

<Canvas>
  <Story name="Small" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          start-of-week="1"
          size="small"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### With value

<Canvas>
  <Story name="With value" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          value="01/01/2023"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Open at date

<Canvas>
  <Story name="Open at date" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          helper-text="The calendar will open showing 25th December 2023, if not date selected"
          open-at-date="2023-12-25"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Alternative dialog label

<Canvas>
  <Story name="Alternative dialog label" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          new-dialog-label="true"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Start of week - Sunday

<Canvas>
  <Story name="Start of week - Sunday" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          start-of-week="0"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### With max and min set

<Canvas>
  <Story name="With max and min set" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <script>
          var dp = document.querySelector("#min-max-dp");
          dp.max = new Date("28 September 2024");
        </script>
        <ic-date-picker
          label="When would you like to collect your coffee?"
          id="min-max-dp"
          min="2023-07-28"
          date-format="MM/DD/YYYY"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Only dates in July 2023

<Canvas>
  <Story name="Only dates in July 2023" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          helper-text="Choose a date between the 1st and 31st of July"
          min="2023-07-01"
          max="2023-07-31"
          show-picker-today-button="false"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Days outside month hidden

<Canvas>
  <Story name="Days outside month hidden" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          show-days-outside-month="false"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Working weekdays only

<Canvas>
  <Story name="Working weekdays only" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <script>
          document.querySelector("#weekday-picker").disableDays = [0, 6];
        </script>
        <ic-date-picker
          id="weekday-picker"
          label="When would you like to collect your coffee?"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Disabled dates (until-now)

<Canvas>
  <Story
    name="Disabled dates (until-now)"
    parameters={{ loki: { skip: true } }}
  >
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          disable-until-now
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Disabled dates (from-now)

<Canvas>
  <Story name="Disabled dates (from-now)" parameters={{ loki: { skip: true } }}>
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          disable-from-now
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>

### Today & clear buttons hidden

<Canvas>
  <Story
    name="Today & clear buttons hidden"
    parameters={{ loki: { skip: true } }}
  >
    {(args) =>
      html`
        <ic-date-picker
          label="When would you like to collect your coffee?"
          show-picker-today-button="false"
          show-picker-clear-button="false"
        ></ic-date-picker>
      `
    }
  </Story>
</Canvas>
