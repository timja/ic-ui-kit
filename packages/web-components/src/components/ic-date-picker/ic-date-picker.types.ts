export type IcDatePickerSizes = "default" | "large" | "small";
