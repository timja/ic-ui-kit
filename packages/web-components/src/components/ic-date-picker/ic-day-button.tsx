/* istanbul ignore file */
import { h, FunctionalComponent } from "@stencil/core";
// import { DAY_NAMES, MONTH_NAMES } from "../../utils/helpers";
import {
  IcDayNames,
  IcDateInputMonths,
  stringEnumToArray,
} from "../../utils/helpers";
import { IcDatePickerSizes } from "./ic-date-picker.types";

export type DayButtonProps = {
  size: IcDatePickerSizes;
  focussed: boolean;
  today: boolean;
  day: Date;
  monthInView: number;
  onSelectDay: (day: Date) => void;
  selected: boolean;
  focussedDayRef?: (element: HTMLButtonElement) => void;
  // disabled: boolean
  inRange: boolean;
  showDaysOutsideMonth: boolean;
  disabledDay: boolean;
  // isSelected: boolean
  // dateFormatter: Intl.DateTimeFormat
  // onDaySelect: (event: MouseEvent, day: Date) => void
  // onKeyboardNavigation: (event: KeyboardEvent) => void
  // focusedDayRef?: (element: HTMLElement) => void
};

export const DayButton: FunctionalComponent<DayButtonProps> = ({
  size,
  focussed,
  today,
  day,
  monthInView,
  onSelectDay,
  selected,
  focussedDayRef,
  inRange,
  showDaysOutsideMonth,
  disabledDay,
  // onDaySelect,
  // onKeyboardNavigation,
  // focusedDayRef,
  // disabled,
  // inRange,
  // isSelected,
  // dateFormatter,
}) => {
  // const isToday = dateMatches(day, today)
  // const isMonth = isEqualMonth(day, focusedDay)
  // const isFocused = isEqual(day, focusedDay)
  // const isOutsideRange = !inRange

  const handleDayClick = (): void => {
    onSelectDay(day);
  };
  const disabled = false;

  const dayNames = stringEnumToArray(IcDayNames);
  const months = stringEnumToArray(IcDateInputMonths);
  const typography = size === "large" ? "subtitle-large" : "caption";
  const outsideRange = !inRange || disabledDay;
  const outsideMonth = monthInView !== day.getMonth();

  return (
    <button
      class={{
        "day-button": true,
        "outside-month": outsideMonth,
        "outside-range": outsideRange,
        today: today,
        selected: selected,
        focussed: focussed,
      }}
      tabIndex={focussed ? 0 : -1}
      aria-disabled={disabled}
      aria-pressed={selected ? "true" : "false"}
      aria-current={today ? "date" : undefined}
      aria-label={
        !disabled
          ? `Choose ${dayNames[day.getDay()]}, ${day.getDate()} ${
              months[day.getMonth()]
            } ${day.getFullYear()}`
          : undefined
      }
      disabled={outsideRange || (outsideMonth && !showDaysOutsideMonth)}
      onClick={handleDayClick}
      ref={(el: HTMLButtonElement) => {
        if (focussed && el && focussedDayRef) {
          focussedDayRef(el);
        }
      }}
    >
      {(!outsideMonth || (outsideMonth && showDaysOutsideMonth)) && (
        <ic-typography variant={typography}>{day.getDate()}</ic-typography>
      )}
    </button>
  );
};
