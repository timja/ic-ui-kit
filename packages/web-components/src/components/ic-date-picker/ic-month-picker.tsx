/* istanbul ignore file */
import { h, FunctionalComponent } from "@stencil/core";
// import { DAY_NAMES, MONTH_NAMES } from "../../utils/helpers";
// import {
//   IcDateInputMonths,
//   stringEnumToArray,
// } from "../../utils/helpers";
import { IcDatePickerSizes } from "./ic-date-picker.types";

export type MonthPickerProps = {
  size: IcDatePickerSizes;
  focussedMonth: number;
  monthInView: number;
  // hidden: boolean;
  monthNames: string[];
  // day: Date;
  onSelectMonth: (month: number) => void;
  onKeyDown: (ev: KeyboardEvent) => void;
  // focussedMonthRef?: (element: HTMLIcButtonElement) => void;
  // selected: boolean;
  // disabled: boolean
  // inRange: boolean
  // isSelected: boolean
  // dateFormatter: Intl.DateTimeFormat
  // onDaySelect: (event: MouseEvent, day: Date) => void
  // onKeyboardNavigation: (event: KeyboardEvent) => void
  // focusedDayRef?: (element: HTMLElement) => void
};

export const MonthPicker: FunctionalComponent<MonthPickerProps> = ({
  size,
  focussedMonth,
  monthInView,
  onSelectMonth,
  // hidden,
  monthNames,
  onKeyDown,
  // focussedMonthRef?,

  // onDaySelect,
  // onKeyboardNavigation,
  // focusedDayRef,
  // disabled,
  // inRange,
  // isSelected,
  // dateFormatter,
}) => {
  // const isToday = dateMatches(day, today)
  // const isMonth = isEqualMonth(day, focusedDay)
  // const isFocused = isEqual(day, focusedDay)
  // const isOutsideRange = !inRange

  const handleMonthClick = (ev: MouseEvent): void => {
    const button = ev.target as HTMLElement;
    // console.log(button.getAttribute("data-month"));
    // this.monthInView = Number(button.getAttribute("data-month"));
    onSelectMonth(Number(button.getAttribute("data-month")));
  };
  // const disabled = false;

  // const months = stringEnumToArray(IcDateInputMonths);
  const buttonSize = size === "small" ? "small" : "default";

  let focussedMonthEl: HTMLIcButtonElement;

  // console.log(focussedMonthEl)

  setTimeout(() => {
    focussedMonthEl.setFocus(), 100;
  });

  return (
    <div
      class={{
        "month-picker": true,
        // hidden: hidden,
      }}
    >
      {monthNames.map((month, index) => {
        const current = monthInView === index;
        const focussed = focussedMonth === index;
        return (
          <ic-button
            class={{
              "month-button": true,
              selected: current,
            }}
            full-width
            variant={current ? "primary" : "tertiary"}
            data-month={index}
            size={buttonSize}
            tabIndex={focussed ? 0 : -1}
            aria-label={current ? `choose ${month}` : `${month} selected`}
            onClick={handleMonthClick}
            onKeyDown={onKeyDown}
            ref={(el: HTMLIcButtonElement) => {
              if (focussed && el) {
                focussedMonthEl = el;
              }
            }}
          >
            {month}
          </ic-button>
        );
      })}
    </div>
  );
};
