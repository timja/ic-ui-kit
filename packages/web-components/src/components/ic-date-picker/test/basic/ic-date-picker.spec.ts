import { newSpecPage } from "@stencil/core/testing";
import { DatePicker } from "../../ic-date-picker";
// import { waitForTimeout } from "../../testspec.setup";

describe("ic-date-picker", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [DatePicker],
      html: `<ic-date-picker label="Date picker label"></ic-date-picker>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
