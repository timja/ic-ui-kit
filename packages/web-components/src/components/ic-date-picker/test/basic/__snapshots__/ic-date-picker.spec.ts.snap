// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-date-picker should render 1`] = `
<ic-date-picker class="default" label="Date picker label">
  <mock:shadow-root>
    <div class="date-input-container">
      <ic-date-input label="Date picker label" showcalendarbutton="" showclearbutton="" value=""></ic-date-input>
    </div>
    <div aria-label="choose date" aria-modal="true" class="calendar-container" role="dialog">
      <span aria-live="polite" class="sr-only">
        in view. Please select a day.
      </span>
      <div class="month-year-nav-container">
        <div class="month-year-nav">
          <div aria-hidden="true">
            <ic-button aria-hidden="true" class="flip" disabletooltip="" id="previous-month-button" size="small" tabindex="-1" variant="icon">
              svg
            </ic-button>
          </div>
          <span aria-hidden="true" id="select-month-hint">
            undefined undefined selected. Click to open month picker view or use the left and right arrow keys to change month.
          </span>
          <ic-button aria-describedby="select-month-hint" aria-label="select month" aria-pressed="false" class="month-picker-button" full-width="true" id="month-picker-button" size="small" variant="tertiary"></ic-button>
          <div aria-hidden="true">
            <ic-button aria-hidden="true" disabletooltip="" id="next-month-button" size="small" tabindex="-1" variant="icon">
              svg
            </ic-button>
          </div>
        </div>
        <div class="month-year-nav">
          <div aria-hidden="true">
            <ic-button aria-hidden="true" class="flip" disabletooltip="" id="previous-year-button" size="small" tabindex="-1" variant="icon">
              svg
            </ic-button>
          </div>
          <span aria-hidden="true" id="select-year-hint">
            undefined selected. Click to open year picker view or use the left and right arrow keys to change year.
          </span>
          <ic-button aria-describedby="select-year-hint" aria-label="select year" aria-pressed="false" class="year-picker-button" full-width="true" id="year-picker-button" size="small" variant="tertiary"></ic-button>
          <div aria-hidden="true">
            <ic-button aria-hidden="true" disabletooltip="" id="next-year-button" size="small" tabindex="-1" variant="icon">
              svg
            </ic-button>
          </div>
        </div>
      </div>
      <div aria-label="undefined undefined" class="calendar">
        <div class="weekdays">
          <div class="calendar-day-header">
            <ic-typography variant="caption">
              Mon
            </ic-typography>
          </div>
          <div class="calendar-day-header">
            <ic-typography variant="caption">
              Tue
            </ic-typography>
          </div>
          <div class="calendar-day-header">
            <ic-typography variant="caption">
              Wed
            </ic-typography>
          </div>
          <div class="calendar-day-header">
            <ic-typography variant="caption">
              Thu
            </ic-typography>
          </div>
          <div class="calendar-day-header">
            <ic-typography variant="caption">
              Fri
            </ic-typography>
          </div>
          <div class="calendar-day-header">
            <ic-typography variant="caption">
              Sat
            </ic-typography>
          </div>
          <div class="calendar-day-header">
            <ic-typography variant="caption">
              Sun
            </ic-typography>
          </div>
        </div>
        <div class="calendar-days-container"></div>
      </div>
      <div class="bottom-buttons">
        <ic-button aria-label="go to today" size="small" variant="tertiary">
          Go to today
        </ic-button>
        <ic-button aria-label="clear selected date" disabled="" id="clear-button" size="small" variant="tertiary">
          Clear
        </ic-button>
      </div>
    </div>
  </mock:shadow-root>
</ic-date-picker>
`;
